Alerting: Panic after failing to create an Alertmanager for an org (duplicate metrics collector registration attempted)

### What happened?

Failing to create an Alertmanager for an org results in a panic due to the metrics collector being registered more than once.

The MultiOrg Alertmanager holds Alertmanagers for the different organizations in an `orgID -> Alertmanager` map. The factory function for Alertmanagers creates some metrics, which then uses to create the Alertmanager.

https://github.com/grafana/grafana/blob/e43ddd516deb0ed720e9d607df7a4c60205ad9b3/pkg/services/ngalert/notifier/multiorg_alertmanager.go#L154-L159

The problem is that when `NewAlertmanager()` returns an error, metrics have already been registered. The next time we attempt to create an Alertmanager for the same org, we also attempt to register the same metrics again, causing a panic.


### What did you expect to happen?

No panics when failing to create an Alertmanager for an org.

### Did this work before?

-

### How do we reproduce it?

1. Return an error from the factory function (e.g. instead of [calling `NewAlertmanager()`](https://github.com/grafana/grafana/blob/e43ddd516deb0ed720e9d607df7a4c60205ad9b3/pkg/services/ngalert/notifier/multiorg_alertmanager.go#L158), return `nil, fmt.Errorf("test")`)
2. Start Grafana
3. After failing to create the Alertmanager and trying again, Grafana will crash


### Is the bug inside a dashboard panel?

_No response_

### Environment (with versions)?




### Grafana platform?

None

### Datasource(s)?

_No response_

```
panic: duplicate metrics collector registration attempted

goroutine 311 [running]:
github.com/prometheus/client_golang/prometheus.(*wrappingRegisterer).MustRegister(0xc00a61f0b0, {0xc007158980?, 0x0?, 0x0?})
	github.com/prometheus/client_golang@v1.19.1/prometheus/wrap.go:104 +0x12b
github.com/prometheus/alertmanager/api/metrics.NewAlerts({0x2743d00, 0xc00a61f0b0})
	github.com/prometheus/alertmanager@v0.27.0/api/metrics/metrics.go:39 +0x4c9
github.com/grafana/grafana/pkg/services/ngalert/metrics.NewAlertmanagerMetrics({0x2742880, 0xc004d17130})
	github.com/grafana/grafana/pkg/services/ngalert/metrics/alertmanager.go:21 +0xe5
github.com/grafana/grafana/pkg/services/ngalert/notifier.NewMultiOrgAlertmanager.func1({0x27649c0, 0xc00640a0a0}, 0x47)
	github.com/grafana/grafana/pkg/services/ngalert/notifier/multiorg_alertmanager.go:156 +0x85
github.com/grafana/grafana/pkg/services/ngalert/notifier.(*MultiOrgAlertmanager).SyncAlertmanagersForOrgs(0xc000e3c460, {0x27649c0, 0xc00640a0a0}, {0xc00c188800, 0x8c, 0x1?})
	github.com/grafana/grafana/pkg/services/ngalert/notifier/multiorg_alertmanager.go:304 +0x322
github.com/grafana/grafana/pkg/services/ngalert/notifier.(*MultiOrgAlertmanager).LoadAndSyncAlertmanagersForOrgs(0xc000e3c460, {0x27649c0, 0xc00640a0a0})
	github.com/grafana/grafana/pkg/services/ngalert/notifier/multiorg_alertmanager.go:260 +0xce
github.com/grafana/grafana/pkg/services/ngalert/notifier.(*MultiOrgAlertmanager).Run(0xc000e3c460, {0x27649c0, 0xc00640a0a0})
	github.com/grafana/grafana/pkg/services/ngalert/notifier/multiorg_alertmanager.go:243 +0xeb
github.com/grafana/grafana/pkg/services/ngalert.(*AlertNG).Run.func1()
	github.com/grafana/grafana/pkg/services/ngalert/ngalert.go:510 +0x26
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78 +0x56
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 7891
	golang.org/x/sync@v0.8.0/errgroup/errgroup.go:75 +0x96
```

--- a/pkg/services/ngalert/metrics/alertmanager.go
+++ b/pkg/services/ngalert/metrics/alertmanager.go
@@ -60,7 +60,11 @@ func NewAlertmanagerConfigMetrics(r prometheus.Registerer) *AlertmanagerConfigMe
 		}),
 	}
 	if r != nil {
-		r.MustRegister(m.ConfigHash, m.ConfigSizeBytes, m.Matchers, m.MatchRE, m.Match, m.ObjectMatchers)
+		for _, c := range []prometheus.Collector{m.ConfigHash, m.ConfigSizeBytes, m.Matchers, m.MatchRE, m.Match, m.ObjectMatchers} {
+			if err := r.Register(c); err != nil {
+				// log err
+			}
+		}
 	}
 	return m
 }

--- a/pkg/services/ngalert/metrics/alertmanager.go
+++ b/pkg/services/ngalert/metrics/alertmanager.go
@@ -3,6 +3,8 @@ package metrics
 import (
 	"fmt"
 
+	"github.com/grafana/grafana/pkg/infra/log"
+
 	"github.com/prometheus/alertmanager/api/metrics"
 	"github.com/prometheus/client_golang/prometheus"
 )
@@ -14,12 +16,12 @@ type Alertmanager struct {
 }
 
 // NewAlertmanagerMetrics creates a set of metrics for the Alertmanager of each organization.
-func NewAlertmanagerMetrics(r prometheus.Registerer) *Alertmanager {
+func NewAlertmanagerMetrics(r prometheus.Registerer, l log.Logger) *Alertmanager {
 	other := prometheus.WrapRegistererWithPrefix(fmt.Sprintf("%s_%s_", Namespace, Subsystem), r)
 	return &Alertmanager{
 		Registerer:                r,
 		Alerts:                    metrics.NewAlerts(other),
-		AlertmanagerConfigMetrics: NewAlertmanagerConfigMetrics(r),
+		AlertmanagerConfigMetrics: NewAlertmanagerConfigMetrics(r, l),
 	}
 }
 
@@ -32,7 +34,7 @@ type AlertmanagerConfigMetrics struct {
 	ObjectMatchers  prometheus.Gauge
 }
 
-func NewAlertmanagerConfigMetrics(r prometheus.Registerer) *AlertmanagerConfigMetrics {
+func NewAlertmanagerConfigMetrics(r prometheus.Registerer, l log.Logger) *AlertmanagerConfigMetrics {
 	m := &AlertmanagerConfigMetrics{
 		ConfigHash: prometheus.NewGaugeVec(prometheus.GaugeOpts{
 			Name: "alertmanager_config_hash",
@@ -61,8 +63,11 @@ func NewAlertmanagerConfigMetrics(r prometheus.Registerer) *AlertmanagerConfigMe
 	}
 	if r != nil {
 		for _, c := range []prometheus.Collector{m.ConfigHash, m.ConfigSizeBytes, m.Matchers, m.MatchRE, m.Match, m.ObjectMatchers} {
+			// simpler than handling prometheus.AlreadyRegisteredError and replacing collectors
+			r.Unregister(c)
+
 			if err := r.Register(c); err != nil {
-				// log err
+				l.Error("Error registering prometheus collector for new alertmanager", "error", err)
 			}
 		}
 	}

--- a/pkg/services/ngalert/metrics/alertmanager_test.go
+++ b/pkg/services/ngalert/metrics/alertmanager_test.go
@@ -1,0 +1,34 @@
+package metrics
+
+import (
+	"testing"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_NewAlertmanagerConfigMetrics(t *testing.T) {
+	t.Run("Successfully registers collectors despite previous call", func(t *testing.T) {
+		r := prometheus.NewRegistry()
+		l := log.NewNopLogger()
+
+		require.NotPanics(t, func() {
+			for i := 0; i < 3; i++ {
+				m := NewAlertmanagerConfigMetrics(r, l)
+				m.ConfigHash.WithLabelValues("test").Set(1)
+				m.Matchers.Set(1)
+				m.MatchRE.Set(1)
+				m.Match.Set(1)
+				m.ObjectMatchers.Set(1)
+
+				mf, err := r.Gather()
+				require.NoError(t, err)
+
+				for j := 0; j < len(mf); j++ {
+					require.Equal(t, float64(1), mf[j].GetMetric()[0].GetGauge().GetValue())
+				}
+			}
+		})
+	})
+}

--- a/pkg/services/ngalert/notifier/alertmanager_test.go
+++ b/pkg/services/ngalert/notifier/alertmanager_test.go
@@ -31,7 +31,9 @@ func setupAMTest(t *testing.T) *alertmanager {
 		AppURL:   "http://localhost:9093",
 	}
 
-	m := metrics.NewAlertmanagerMetrics(prometheus.NewRegistry())
+	l := log.New("alertmanager-test")
+
+	m := metrics.NewAlertmanagerMetrics(prometheus.NewRegistry(), l)
 	sqlStore := db.InitTestDB(t)
 	s := &store.DBstore{
 		Cfg: setting.UnifiedAlertingSettings{
@@ -39,7 +41,7 @@ func setupAMTest(t *testing.T) *alertmanager {
 			DefaultRuleEvaluationInterval: time.Minute,
 		},
 		SQLStore:         sqlStore,
-		Logger:           log.New("alertmanager-test"),
+		Logger:           l,
 		DashboardService: dashboards.NewFakeDashboardService(t),
 	}
 

--- a/pkg/services/ngalert/notifier/multiorg_alertmanager.go
+++ b/pkg/services/ngalert/notifier/multiorg_alertmanager.go
@@ -153,7 +153,7 @@ func NewMultiOrgAlertmanager(
 
 	// Set up the default per tenant Alertmanager factory.
 	moa.factory = func(ctx context.Context, orgID int64) (Alertmanager, error) {
-		m := metrics.NewAlertmanagerMetrics(moa.metrics.GetOrCreateOrgRegistry(orgID))
+		m := metrics.NewAlertmanagerMetrics(moa.metrics.GetOrCreateOrgRegistry(orgID), l)
 		stateStore := NewFileStore(orgID, kvStore)
 		return NewAlertmanager(ctx, orgID, moa.settings, moa.configStore, stateStore, moa.peer, moa.decryptFn, moa.ns, m, featureManager.IsEnabled(ctx, featuremgmt.FlagAlertingSimplifiedRouting))
 	}
@@ -211,7 +211,6 @@ func (moa *MultiOrgAlertmanager) setupClustering(cfg *setting.Cfg) error {
 			true,
 			cfg.UnifiedAlerting.HALabel,
 		)
-
 		if err != nil {
 			return fmt.Errorf("unable to initialize gossip mesh: %w", err)
 		}
@@ -358,7 +357,8 @@ func (moa *MultiOrgAlertmanager) SyncAlertmanagersForOrgs(ctx context.Context, o
 // active organizations. The original intention for this was the cleanup deleted orgs, that have had their states
 // saved to the kvstore after deletion on instance shutdown.
 func (moa *MultiOrgAlertmanager) cleanupOrphanLocalOrgState(ctx context.Context,
-	activeOrganizations map[int64]struct{}) {
+	activeOrganizations map[int64]struct{},
+) {
 	storedFiles := []string{NotificationLogFilename, SilencesFilename}
 	for _, fileName := range storedFiles {
 		keys, err := moa.kvStore.Keys(ctx, kvstore.AllOrganizations, KVNamespace, fileName)

